compiler/damlc/lib/DA/Cli/Damlc/Command/MultiIde/Types.hs

```diff
@@ -0,0 +1,182 @@
+type InitParams = LSP.InitializeParams
+type InitParamsVar = MVar InitParams
+
+-- Stores the mapping of server created progress tokens and their IDEs home to their prefixed counterpart 
+-- We need this so we know which of the Progress messages from the server need to be prefixed
+--   since ProgressTokens created by the client are not prefixed.
+type ServerCreatedProgressTokens = Map.Map (LSP.ProgressToken, FilePath) LSP.ProgressToken
+type ServerCreatedProgressTokensVar = MVar ServerCreatedProgressTokens
+
+-- Maps a packages unit id to its source file path, for all packages listed in a multi-package.yaml
+type MultiPackageYamlMapping = Map.Map String FilePath
+
+data MultiIdeState = MultiIdeState
+  { fromClientMethodTrackerVar :: MethodTrackerVar 'LSP.FromClient
+    -- ^ The client will track its own IDs to ensure they're unique, so no worries about collisions
+  , fromServerMethodTrackerVar :: MethodTrackerVar 'LSP.FromServer
+    -- ^ We will prefix LspIds before they get here based on their SubIDE messageIdPrefix, to avoid collisions
+  , serverCreatedProgressTokensVar :: ServerCreatedProgressTokensVar
+  , subIDEsVar :: SubIDEsVar
+  , initParamsVar :: InitParamsVar
+  , toClientChan :: TChan BSL.ByteString
+  , multiPackageMapping :: MultiPackageYamlMapping
+  }
+
+newMultiIdeState :: MultiPackageYamlMapping -> IO MultiIdeState
+newMultiIdeState multiPackageMapping = do
+  (fromClientMethodTrackerVar :: MethodTrackerVar 'LSP.FromClient) <- newTVarIO IM.emptyIxMap
+  (fromServerMethodTrackerVar :: MethodTrackerVar 'LSP.FromServer) <- newTVarIO IM.emptyIxMap
+  serverCreatedProgressTokensVar <- newMVar @ServerCreatedProgressTokens mempty
+  subIDEsVar <- newTMVarIO @SubIDEs mempty
+  initParamsVar <- newEmptyMVar @InitParams
+  toClientChan <- atomically newTChan
+  pure MultiIdeState {..}
+
+-- Forwarding
+
+{-
+Types of behaviour we want:
+
+Regularly handling by a single IDE - works for requests and notifications
+  e.g. TextDocumentDidOpen
+Ignore it
+  e.g. Initialize
+Forward a notification to all IDEs
+  e.g. workspace folders changed, exit
+Forward a request to all IDEs and somehow combine the result
+  e.g.
+    symbol lookup -> combine monoidically
+    shutdown -> response is empty, so identity after all responses
+  This is the hard one as we need some way to define the combination logic
+    which will ideally wait for all IDEs to reply to the request and apply this function over the (possibly failing) result
+  This mostly covers FromClient requests that we can't pull a filepath from
+
+  Previously thought we would need this more, now we only really use it for shutdown - ensuring all SubIdes shutdown before replying.
+  We'll keep it in though since we'll likely get more capabilities supported when we upgrade ghc/move to HLS
+-}
+
+-- TODO: Consider splitting this into one data type for request and one for notification
+-- rather than reusing the Single constructor over both and restricting via types
```

Comment:
This will happen when I make the type family specified in TODO above

Reply:
Grand, its on you to verify the TODO and remove it then when you're done ;)